src/Cli/dotnet/PerformanceLogEventSource.cs

```diff
@@ -0,0 +1,341 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Tools.Common;
+using RuntimeEnvironment = Microsoft.DotNet.Cli.Utils.RuntimeEnvironment;
+
+namespace Microsoft.DotNet.Cli
+{
+    [EventSource(Name = "Microsoft-Dotnet-CLI-Performance")]
+    internal sealed class PerformanceLogEventSource : EventSource
+    {
+        internal static PerformanceLogEventSource Log = new PerformanceLogEventSource();
```

Comment:
I don't know the implementation detail. Will a static PerformanceLogEventSource not cause a race condition in unit tests running in parallel? There might be multiple tests want to init the object at the sometime. And some might want to deconstruct it.

Reply:
If there are multiple tests that happen to run this code inside the same process, then the first caller will execute the static constructor, while the rest will wait behind a lock.  If the tests are all running in separate processes, then execution of the static constructors won't impact each other.  What could have an impact is that we clean-up old perf log files.  We might need to find a way to address this if we want to do any considerable execution during tests with this feature on.  One way to solve the problem might be to increase the number of logs that we keep by environment variable, so that the test harness can ensure that no logs get deleted if tests are looking for them.

---
In reply to: [505757906](https://github.com/wli3/sdk/pull/1#discussion_r505757906) [](ancestors = 505757906)

Reply:
> then the first caller will execute the static constructor, while the rest will wait behind a lock

The lock is in EventSource constructor? 

> What could have an impact is that we clean-up old perf log files

During unit test we should only have the TestEventListener that only write things to memory. Or you are saying init PerformanceLogEventListener is unavoidable?

Reply:
The lock is in the runtime.  The runtime guarantees that the static constructor won't get executed multiple times.

The strategy you describe - using an in-memory listener, could work.  But I think we would want to make sure that creation of logs and clean-up of logs works.  One strategy that we could use is to redirect the storage location to something that can be controlled by the test.  That way, we ensure that two tests running in parallel don't impact each other.  How does that sound?

Reply:
> make sure that creation of logs and clean-up of logs works

Where to store and how to store is a separate concern of what to log. We should test it separately. Another problem, if we one day makes it enable by default, we also need to solve the issue of PerformanceLogEventListener being enabled for every single functional test. It is a lot of wasted file at least. By current design, there should not be a race condition right?